plugins/logs/mask.go

```diff
@@ -0,0 +1,234 @@
+package logs
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/open-policy-agent/opa/util"
+)
+
+type maskOP string
+
+const (
+	maskOPRemove maskOP = "remove"
+	maskOPUpsert maskOP = "upsert"
+)
+
+type maskRule struct {
+	OP             maskOP      `json:"op"`
+	Path           string      `json:"path"`
+	Value          interface{} `json:"value"`
+	escapedParts   []string
+	unEscapedParts []string
+}
+
+func (r maskRule) String() string {
+	return "/" + strings.Join(r.escapedParts, "/")
+}
+
+func newMaskRule(op maskOP, path string, val interface{}) (maskRule, error) {
+	rule := maskRule{}
+
+	if len(path) == 0 {
+		return rule, fmt.Errorf("mask must be non-empty")
+	} else if path[0] != '/' {
+		return rule, fmt.Errorf("mask must be slash-prefixed")
+	}
+
+	parts := strings.Split(path[1:], "/")
+
+	if parts[0] != "input" && parts[0] != "result" {
+		return rule, fmt.Errorf("mask prefix not allowed")
+	}
+
+	unEscapedParts := make([]string, len(parts))
+	escapedParts := make([]string, len(parts))
+	for i := range parts {
+		var err error
+		unEscapedParts[i], err = url.QueryUnescape(parts[i])
+		if err != nil {
+			return rule, err
+		}
+
+		escapedParts[i] = url.QueryEscape(parts[i])
+	}
+
+	rule = maskRule{
+		OP:             op,
+		Path:           path,
+		Value:          val,
+		unEscapedParts: unEscapedParts,
+		escapedParts:   escapedParts,
+	}
+
+	return rule, nil
+}
+
+func (r maskRule) Erase(event *EventV1) {
+
+	if len(r.escapedParts) == 1 {
+		switch r.escapedParts[0] {
+		case "input":
+			switch r.OP {
+			case maskOPRemove:
+				event.Input = nil
+				event.Erased = append(event.Erased, r.String())
+
+			case maskOPUpsert:
+				event.Input = &r.Value
+				event.Masked = append(event.Masked, r.String())
+
+			default:
+				// TODO(dkiser) Do we want to panic instead of return errors here?
+				panic(fmt.Sprintf("illegal mask op value: %s", r.OP))
```

Comment:
Lets have this return an error up to the logger plugin, the plugin may or may not be able to do anything other than log an error but its better than a panic. That goes for all of the panics in here.